x/qgb/README.md

```diff
@@ -1,50 +1,320 @@
+
+To check if the earliest attestation nonce is defined in store, use the [`CheckEarliestAvailableAttestationNonce(...)`](https://github.com/celestiaorg/celestia-app/blob/9bf0cf1dd9ce31a3fecb51310c3913820b21a8c2/x/qgb/keeper/keeper_attestation.go#L81-L87) method.
 
-The data commitment window is defined as a governance parameter that can be changed. Currently, it is set to 400.
+## State Transitions
 
-### Panics
+### End Block
+
+During the `EndBlock` step, we're executing the [logic](https://github.com/celestiaorg/celestia-app/blob/9bf0cf1dd9ce31a3fecb51310c3913820b21a8c2/x/qgb/abci.go#L28-L35) that handles the generating new valsets, new data commitments, and prunes when needed.
+
+### Valset handler
+
+A new valset is generated by the valset [`handler`](https://github.com/celestiaorg/celestia-app/blob/9bf0cf1dd9ce31a3fecb51310c3913820b21a8c2/x/qgb/abci.go#L84-L135) in the following cases:
+
+#### No valset in store
+
+When `EndBlock` is executed straight after genesis, the store doesn't have any valset created yet. So, it [checks](https://github.com/celestiaorg/celestia-app/blob/9bf0cf1dd9ce31a3fecb51310c3913820b21a8c2/x/qgb/abci.go#L85-L93) wether there is any valset in store, and [generates](https://github.com/celestiaorg/celestia-app/blob/9bf0cf1dd9ce31a3fecb51310c3913820b21a8c2/x/qgb/abci.go#L123-L134) a new one afterwards representing the initial validator set.
+
+#### Validator starts unbonding
+
+When a validator starts unbonding, the `LatestUnbondingBlockHeight` gets updated in a hook with the current block number. Then, inside `EndBlock`, we [check](https://github.com/celestiaorg/celestia-app/blob/9bf0cf1dd9ce31a3fecb51310c3913820b21a8c2/x/qgb/abci.go#L123) whether the `LatestUnbondingBlockHeight` corresponds to the current block height. If so, then we generate a new valset that doesn't contain that unbonding validator.
+
+#### Significant power change
+
+The third scenario where a valset gets created is when there is a significant power change. As stated above, valsets contain an `evmAddress -> power` mapping for the validator sets they represent. When a [significant power change](https://github.com/celestiaorg/celestia-app/blob/9bf0cf1dd9ce31a3fecb51310c3913820b21a8c2/x/qgb/abci.go#L99-L120) happens, a new valset gets created. The significant power threshold is defined by the [`SignificantPowerDifferenceThreshold`](https://github.com/celestiaorg/celestia-app/blob/9bf0cf1dd9ce31a3fecb51310c3913820b21a8c2/x/qgb/abci.go#L15-L18) constant.
+
+A significant power change can happen if a validator's delegation got reduced or increased significantly, or the powers of multiple validators changed in a way that the whole validator set variation is higher than the threshold. This calculus is done inside the [`PowerDiff(...)`](https://github.com/celestiaorg/celestia-app/blob/9bf0cf1dd9ce31a3fecb51310c3913820b21a8c2/x/qgb/types/validator.go#L100-L140) method.
```

Comment:
No upper bound. And no rationale, inherited from the gravity bridge

Reply:
I guess 5% errs on the side of caution but the upper bound might be something like 33% or 1/3. I haven't spent too much time thinking about the attack vectors here but it might be good to eventually write them up as a form of rationale

Reply:
I will open an issue to think on an upper bound for this value, thanks for pointing this out

Reply:
https://github.com/celestiaorg/celestia-app/issues/1920